compatibility_spark-3/src/main/scala/za/co/absa/hyperdrive/compatibility/impl/DeltaCDCToSCD2Writer.scala

```diff
@@ -0,0 +1,245 @@
+  private val logger = LoggerFactory.getLogger(this.getClass)
+
+  private val CHECKPOINT_LOCATION = "checkpointLocation"
+
+  private val STRING_SEPARATOR = "#$@"
+  private val START_DATE_COLUMN = "_start_date"
+  private val END_DATE_COLUMN = "_end_date"
+  private val IS_CURRENT_COLUMN = "_is_current"
+  private val MERGE_KEY_COLUMN = "_mergeKey"
+  private val SORT_FIELD_PREFIX = "_tmp_hyperdrive_"
+  val OLD_DATA_PREFIX = "_oldData"
+  val NEW_DATA_PREFIX = "_newData"
+
+  if (configuration.precombineColumnsCustomOrder.values.flatten.toSeq.contains(STRING_SEPARATOR)) {
+    throw new IllegalArgumentException(s"Precombine columns custom order cannot contain string separator: $STRING_SEPARATOR")
+  }
+
+  override def write(dataFrame: DataFrame): StreamingQuery = {
+    dataFrame.writeStream
+      .trigger(configuration.trigger)
+      .outputMode(OutputMode.Append())
+      .option(CHECKPOINT_LOCATION, configuration.checkpointLocation)
+      .options(configuration.extraConfOptions)
+      .foreachBatch((df: DataFrame, batchId: Long) => {
+        if (!DeltaTable.isDeltaTable(df.sparkSession, configuration.destination)) {
+          if (isDirEmptyOrDoesNotExist(df.sparkSession, configuration.destination)) {
+            logger.info(s"Destination: ${configuration.destination} is not a delta table. Creating new delta table.")
+            df.sparkSession
+              .createDataFrame(
+                df.sparkSession.sparkContext.emptyRDD[Row],
+                df.schema
+                  .add(START_DATE_COLUMN, TimestampType, nullable = false)
+                  .add(END_DATE_COLUMN, TimestampType, nullable = true)
+                  .add(IS_CURRENT_COLUMN, BooleanType, nullable = false)
+              )
+              .write
+              .format("delta")
+              .mode(SaveMode.Overwrite)
+              .option("overwriteSchema", "true")
+              .partitionBy(configuration.partitionColumns: _*)
+              .save(configuration.destination)
+          } else {
+            throw new IllegalArgumentException(s"Could not create new delta table. Directory ${configuration.destination} is not empty!")
+          }
+        }
+        logger.info(s"Writing batchId: $batchId")
+
+        val deltaTable = DeltaTable.forPath(configuration.destination)
+
+        val uniqueChangesForEachKeyAndTimestamp = removeDuplicates(df)
+        val previousEvents = getPreviousEvents(deltaTable, uniqueChangesForEachKeyAndTimestamp)
+        val nextEvents = getNextEvents(deltaTable, uniqueChangesForEachKeyAndTimestamp)
+
+        val union = previousEvents.union(nextEvents).distinct().unionAll(
+          uniqueChangesForEachKeyAndTimestamp
+            .withColumn(START_DATE_COLUMN, col(configuration.timestampColumn))
+            .withColumn(END_DATE_COLUMN, lit(null))
+            .withColumn(IS_CURRENT_COLUMN, lit(false))
+            .withColumn(MERGE_KEY_COLUMN, lit(null))
```

Comment:
`MERGE_KEY_COLUMN` seems to be only needed to check if it's null or not. I suggest changing it to a boolean column, something like `IS_OLD_DATA`.

Reply:
Done